source/common/http/conn_manager_impl.cc

```diff
@@ -1058,6 +1058,13 @@ void ConnectionManagerImpl::ActiveStream::encodeHeaders(ActiveStreamEncoderFilte
   }
 }
 
+HeaderMap& ConnectionManagerImpl::ActiveStream::addEncodedTrailers() {
+  if (!response_trailers_) {
```

Comment:
this whole interaction ends up becoming somewhat complicated so i wanted to put up the PR before spending too much time on the edge cases:

the `make_unique` allows calling this before `encodeTrailers` is called, but when `encodeTrailers` is called later on it will overwrite this value by moving the upstream trailers into this variable. This means that right now this should only be called in `encodeData` if `encodeTrailers` won't ever be called. When called in `encodeTrailers` it should correctly modify the right map. 

This similarly applies to `encodeHeaders`: if there are upstream trailers they will overwrite anything that was added in `encodeHeaders`

Reply:
I commented above. IMO this is a real filter edge case, and I think I would just go around all of this and pass the new trailers by move semantics, which then overwrite. Thoughts?

Reply:
I think that makes sense and works with our use case. Seems cleaner than to have to reason about how it may interact with existing trailers.